### ISSUE TITLE
[5.0.0rc12] can't decide between rocm and cuda accelerator

### ISSUE THREAD

**Issue by lahwaacz** [open-mpi/ompi#11878]
I built the 5.0.0rc12 release with these flags
```
  Configure command line: '--prefix=/usr' '--enable-builtin-atomics'
                          '--enable-memchecker' '--enable-mpi-fortran=all'
                          '--enable-pretty-print-stacktrace'
                          '--libdir=/usr/lib' '--sysconfdir=/etc/openmpi'
                          '--with-ucx=/usr' '--with-cuda=/opt/cuda'
                          '--with-cuda-libdir=/usr/lib'
                          '--with-rocm=/opt/rocm' '--with-ofi=/usr'
                          '--with-hwloc=external' '--with-libevent=external'
                          '--with-pmix=external' '--with-valgrind'
```
and installed on a system which has one AMD GPU card. The full ROCm stack is installed in `/opt/rocm`, CUDA is not installed and `nvidia-utils` is installed only to provide `/usr/lib/libcuda.so.1` that is required for `/usr/lib/libopen-pal.so` (see https://github.com/open-mpi/ompi/issues/11877#issuecomment-1694665726). The operating system is Arch Linux.

But still, running an MPI program with `mpirun` reports the following error:
```
--------------------------------------------------------------------------
ERROR: Open MPI only supports one accelerator framework per node. Please pick one of the
initialized accelerator components via command line argument "--mca accelerator <component>".
The following accelerators initialized successfully: rocm, cuda.
The program will abort to allow you to resolve the accelerator conflict.


--------------------------------------------------------------------------
```

Why does OpenMPI not recognize that the only accelerator available on the system is AMD?

**Comment by edgargabriel** [comment]
@lahwaacz the decision that only one accelerator type is supported per node is an internal design decision of the accelerator framework. We did envision that Open MPI might be compiled with multiple accelerator types, but I don't think we had a scenario in mind where a single node might contain the software stack for multiple accelerator types. 

So the question in your scenario here is, if Open MPI has been compiled with both rocm and cuda support as dsos (as discussed on the other ticket), one component should ideally disqualify itself (assuming that the software stack for that accelerator type is not available on the), and the other one could be used.

Note that you can also explicitly disable a component either using an environment variable or a command line argument (e.g. `--mca accelerator ^cuda` if you only want to use the rocm component).

**Comment by lahwaacz** [comment]
@edgargabriel Building all accelerator components as DSOs as you advised in the other ticket is sufficient for our use case, which is a distribution package for Arch Linux where the accelerator components are optional dependencies for openmpi. When a user installs multiple accelerator software stacks, we can leave explicitly disabling one or the other up to them.

I think the documentation could be improved regarding the use cases and configuration, otherwise feel free to close this issue.

**Comment by jsquyres** [comment]
> I think the documentation could be improved regarding the use cases and configuration

@edgargabriel Could you take a whack at this?

**Comment by jsquyres** [comment]
@edgargabriel Is this complete?

**Comment by edgargabriel** [comment]
@jsquyres If I understand it correctly, we will have all accelerator components in 5.0.1 being built as dso's by default, so half of the request is already fulfilled, users will not fact this issue anymore. The second aspect is that one can exclude a component using the  mca runtime options or environment variables. This is however exactly the same for all other frameworks/components, there is nothing specific to the accelerator framework. We do have documentation that explains the mechanism in general. Not sure I see value in adding explicitly a section about how to exclude an accelerator framework component at runtime.

**Comment by lahwaacz** [comment]
@edgargabriel Indeed, I just built openmpi 5.0.2 without the `--enable-mca-dso` configure flag and all accelerator components are built as dynamic components. Actually, I've used `--enable-mca-dso=accelerator_cuda,rcache_gpusm,rcache_rgpusm` since the rc12 release, but now even the `btl_smcuda` mca is built as dso by default. I did not notice this before, because the `--enable-mca-dso` seems to _override_ the default dso selection, i.e. it does not _append_ to the default-selected list.

I would really appreciate if there was a clear documentation regarding this behavior.
